improved audience handling to support client credentials access tokens without aud claims

kschu91 opened the pull request:
This PR fixes #891. 

## Motivation and Context

Some OIDC Issuers, like AWS Cognito, do not define a `aud` claim, when issuing access tokens using the client credentials flow.
oauth2-proxy allows verifying already issued tokens using `--skip-jwt-bearer-tokens`, but when the `aud` claim is missing, oauth2-proxy fails.

> If the principal processing the claim does not identify itself with a value in the "aud" claim when this claim is present, then the JWT MUST be rejected.

_Quote from https://tools.ietf.org/html/rfc7519_

Having that in mind, oauth2-proxy should not fail, when `aud` is missing. But this will lead to a potential security concern as there will be no verificatioin of the audience.

## Description

This PR adds extra options to optionally skip `aud` verification, if `aud` is missing in the bearer token and the possibility to define an  alternative claim to verify against the client id:
- `skip-aud-check-when-missing` (bool)
- `audience-verification-claim` (string)
- `extra-audiences-for-verification` ([]string)

### skip-aud-check-when-missing
If enabled, audience is only verified if the `aud` claim exists in the token. If not, there is no verification of  audience.

This will internally set `SkipClientIDCheck` in the underlying go-oidc library to true.
https://github.com/AOEpeople/oauth2-proxy/blob/69dfbce99a158556010da4d60c1d5864abd86e3a/pkg/validation/options.go#L123

If set, the verification will happen within oauth2-proxy to have more control over the details of token verification:
https://github.com/AOEpeople/oauth2-proxy/blob/69dfbce99a158556010da4d60c1d5864abd86e3a/providers/provider_data.go#L175

### audience-verification-claim
Can be specified along with `skip-aud-check-when-missing` and allows to specify an alternative claim where to find the client id to verify against. For example AWS Cognito has an `client_id` claim in the access tokens. 

If set, `skip-aud-check-when-missing` will implicitly enabled. oauth2-proxy then verifies this claim instead of the `aud`:
https://github.com/AOEpeople/oauth2-proxy/blob/69dfbce99a158556010da4d60c1d5864abd86e3a/providers/provider_data.go#L185

### extra-audiences-for-verification
An optional list of extra audiences that are allowed to pass the verification. Eg. when the token has been issued by another client than the oauth2-proxy client.

This is similar to the `extra-jwt-issuers` option, but we decided to add is separately since `extra-jwt-issuers` looks like a legacy option.
If set, `skip-aud-check-when-missing` will be implicitly enabled. oauth2-proxy then allows it´s client id, and all in this option defined audiences to pass verification.
https://github.com/AOEpeople/oauth2-proxy/blob/69dfbce99a158556010da4d60c1d5864abd86e3a/providers/provider_data.go#L274

## How Has This Been Tested?

We have implemented this feature donwards compatible. If `audience-verification-claim` is not set, nothing changes.

We have tested it using a local setup.

**Please have a look at the concept and if you are fine with that. If so, we can add unit tests and documentation.**

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

NickMeves commented on pkg/apis/options/legacy_options.go:
```diff
@@ -509,6 +509,10 @@ type LegacyProvider struct {
 	AllowedGroups                      []string `flag:"allowed-group" cfg:"allowed_groups"`
 
+	SkipAudCheckWhenMissing       bool     `flag:"skip-aud-check-when-missing" cfg:"skip_aud_check_when_missing"`
+	AudienceVerificationClaim     string   `flag:"audience-verification-claim" cfg:"audience_verification_claim"`
```
I wonder if this is simpler if just setup as `oidc-aud-claim` that matches `oidc-email-claim` and `oidc-groups-claim` we have to support those claims that some providers have non-standard (similar to your issue with AWS Cognito `aud` vs `client_id` (?) I think you said.

NickMeves commented on pkg/apis/options/legacy_options.go:
```diff
@@ -509,6 +509,10 @@ type LegacyProvider struct {
 
+	SkipAudCheckWhenMissing       bool     `flag:"skip-aud-check-when-missing" cfg:"skip_aud_check_when_missing"`
+	AudienceVerificationClaim     string   `flag:"audience-verification-claim" cfg:"audience_verification_claim"`
+	ExtraAudiencesForVerification []string `flag:"extra-audiences-for-verification" cfg:"extra_audiences_for_verification"`
```
Can we split this functionality out from this PR?

And open an Issue to discuss? I'm not sure this is something we want to support I"m not sure I like freeform audiences that aren't explicitly attached to an issuer that matches it.

`extra-jwt-issuers` will likely be replaced by a concept of `BearerProviders` when we have structured configurations. Where any provider that implements the `CreateSessionFromToken` provider interface method can plug in as a Bearer handler.

At the moment people get super confused conflating these `extra-jwt-issuer` default OIDC handling with their main OIDC provider settings.

NickMeves commented on providers/provider_data.go:
```diff
@@ -168,6 +172,30 @@ func (p *ProviderData) buildSessionFromClaims(idToken *oidc.IDToken) (*sessions.
 		return nil, fmt.Errorf("email in id_token (%s) isn't verified", claims.Email)
 	}
 
+	if p.SkipAudCheckWhenMissing {
```
These token -> session conversions in bearer auth are covered here: https://github.com/oauth2-proxy/oauth2-proxy/blob/06808704a3762de73bed1cc92b077420eb93d735/oauthproxy.go#L310

The challenge is we use `go-oidc` library to verify OIDC IDTokens. That library will not validate tokens without a matching `aud` if I recall.

I don't think this gets around that constraint at all, right?

NickMeves replied:
You need to use `SkipClientIDCheck` on the underlying OIDC Verifier if I recall.

kschu91 commented:
@NickMeves I guess you mentioned some good points. So I have completly restrucutred the implementation.

# oidc-audience-claim
As suggested, I have simplified the logic here by just adding one additional config `oidc-audience-claim`. Which per default is set to "aud" and can be changed for each provider respectively.

# oidc-extra-audiences
In addition to that I added a second config, which per default is empty. If set, the audience(s) from this config are also allowed to pass verification.
Unfortunetly this is required when working with different clients and cognito, as cognito does not allow to add additional audiences to the audience claim.
Since you pointed correctly it may not be a good idea to have extra audiences not bound to an issuer, this config is now bound to the respective provider and it´s issuer.

# architectural considerations
I wrapped the IDTokenVerifier from the underlying oidc lib with an internal one. This allows one to add additional verification logic inside oauth2-proxy and ensures that this verification is also compatible with `extra-jwt-issuers` (at least until this gets replaced by anything else).
At the moment I only added the audience verification to the internaly verifier. But from an architectural point of view it may makes sense to also add other verifications which are currently implemented in `provider_data.go` or somewhere else (eg. the email_verified handling).
As last important side note it is important to understand that with that change `SkipClientIDCheck` from the underlying oidc lib  is always set to `false` and the verification will be handled in oauth2-proxy in any case.

# next steps
Please let me know what you think about this. If you still think it makes sense to split ` oidc-audience-claim` and `oidc-extra-audiences` into separate pull requests just let me know. If you are fine with the implementation I will add some documentation and additional tests.

JoelSpeed commented on pkg/apis/options/options.go:
```diff
@@ -2,9 +2,9 @@ package options
 
 import (
 	"crypto"
+	internaloidc "github.com/oauth2-proxy/oauth2-proxy/v7/pkg/oidc"
```
This shouldn't be here, it should be imported inline with the other oauth2-proxy imports below. I think the linter should be complaining about this

JoelSpeed commented on pkg/oidc/verify.go:
```diff
@@ -0,0 +1,74 @@
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/logger"
+)
+
+type IDTokenVerifier struct {
+	*oidc.IDTokenVerifier
+	*IDTokenVerificationOptions
+}
+
+type IDTokenVerificationOptions struct {
+	AudienceClaim  string
+	ClientID       string
+	ExtraAudiences []string
+}
```
Would be good to get some comments on anything that is exported before we merge to explain what these are and what they're used for please

JoelSpeed commented on pkg/oidc/verify.go:
```diff
@@ -0,0 +1,74 @@
+		logger.Printf("verifying provided aud claim %s with value %s against allowed audiences %v",
+			v.AudienceClaim, audienceClaimValue, allowedAudiences)
+		return v.isValidAudience(audienceClaimValue.(string), allowedAudiences)
+	} else {
```
You don't need this else as there is a return at the end of the previous block, you can unindent the code below if you remove it

JoelSpeed commented on pkg/oidc/verify.go:
```diff
@@ -0,0 +1,74 @@
+		if v.AudienceClaim == "aud" {
+			return false, fmt.Errorf("no valid aud claim exists in token")
+		}
+		if defaultAudienceValue, defaultAudienceExists := claims["aud"]; defaultAudienceExists {
+			logger.Printf("falling back to aud claim, as %s claim does not exists", v.AudienceClaim)
+			return v.isValidAudience(defaultAudienceValue.(string), allowedAudiences)
+		} else {
```
If the user specifies the audience claim, should we really fall back to a default that may not be what they had intended?

kschu91 replied:
Cognito for example uses `aud` claim only in the ID Token. All kind of access tokens do not have an `aud` claim. So it is theoretically not required to get it to work with cognito, as long as the access tokens are used.
I thought it behaves more resilient when falling back to the default, as it is possible to use either id or access token. 
But you are right. That is maybe not what people expect when using this configuration. I will drop it out.

JoelSpeed commented on pkg/oidc/verify.go:
```diff
@@ -0,0 +1,74 @@
+}
+
+func (v *IDTokenVerifier) verifyAudience(claims map[string]interface{}) (bool, error) {
+	allowedAudiences := append([]string{v.ClientID}, v.IDTokenVerificationOptions.ExtraAudiences...)
```
Since this doesn't change for the lifetime of the verifier, what about constructing this at the time the verifier is constructed? You could also make it a map to make the lookup easier

kschu91 commented:
@JoelSpeed please review again. I added/changed/implemented your suggestions. Also, I added a couple of tests and added documentation.

kschu91 commented:
@NickMeves @JoelSpeed would you mind giving this a final review and merge it? Let me know if anything else is required.

Thanks!

kschu91 commented:
@NickMeves @JoelSpeed ping

JoelSpeed commented on pkg/oidc/verify.go:
```diff
@@ -0,0 +1,70 @@
+			v.AudienceClaim, audienceClaimValue, v.allowedAudiences)
+		return v.isValidAudience(audienceClaimValue.(string), v.allowedAudiences)
+	}
+	return false, fmt.Errorf("audience claim %s does not exist in claims: %v",
```
I don't think this message is correct given the allowedAudiences you are passing as the second value, what did you intend here?

kschu91 replied:
This error will be thrown if an `AudienceClaim` is set that does not exist in the set of claims given in the token.

kschu91 replied:
Good catch, I´ll fix the second argument and also add a test for it.

JoelSpeed commented on pkg/oidc/verify.go:
```diff
@@ -0,0 +1,70 @@
+type IDTokenVerifier struct {
+	*oidc.IDTokenVerifier
+	*IDTokenVerificationOptions
+	allowedAudiences []string
```
What about making this a `map[string]struct{}`, then there won't be any duplication and the lookup to check if the audience is allowed will be easier and more efficient

JoelSpeed commented on pkg/oidc/verify_test.go:
```diff
@@ -0,0 +1,202 @@
+)
+
+var _ = Describe("Verify", func() {
+	var (
+		ctx context.Context
+	)
+
+	BeforeEach(func() {
+		ctx = context.Background()
+	})
```
I don't think this is necessary, is the context being overriden anywhere?

The following is probably sufficient in the case of a context
```suggestion
	ctx := context.Background()
```

JoelSpeed commented on pkg/oidc/verify_test.go:
```diff
@@ -0,0 +1,202 @@
+			Aud: "1226737",
+		})
+
+		Expect(err).To(Succeed())
```
This would be more idiomatic/readable if it were
```suggestion
		Expect(err).ToNot(HaveOccurred())
```

JoelSpeed commented on pkg/oidc/verify_test.go:
```diff
@@ -0,0 +1,202 @@
+		})
+
+		Expect(err).To(Succeed())
+		Expect(result).ToNot(BeNil())
```
Is the result defined here? Do we care about the result?

JoelSpeed commented on pkg/oidc/verify_test.go:
```diff
@@ -0,0 +1,202 @@
+			Aud: "1226737",
+		})
+
+		Expect(err).To(Succeed())
```
```suggestion
		Expect(err).ToNot(HaveOccurred())
```

JoelSpeed commented on pkg/oidc/verify_test.go:
```diff
@@ -0,0 +1,202 @@
+			ClientId: "1226737",
+		})
+
+		Expect(err).To(Succeed())
```
```suggestion
		Expect(err).ToNot(HaveOccurred())
```

JoelSpeed commented on pkg/oidc/verify_test.go:
```diff
@@ -0,0 +1,202 @@
+			ClientId: "1226737",
+		})
+
+		Expect(err).To(Succeed())
```
```suggestion
		Expect(err).ToNot(HaveOccurred())
```

JoelSpeed reviewed (commented):
You've changed the filemode on a bunch of files in this PR, please revert those file mode changes before we can merge

kschu91 commented on pkg/validation/options.go:
```diff
@@ -369,9 +382,10 @@ func parseJwtIssuers(issuers []string, msgs []string) ([]jwtIssuer, []string) {
 // newVerifierFromJwtIssuer takes in issuer information in jwtIssuer info and returns
 // a verifier for that issuer.
-func newVerifierFromJwtIssuer(jwtIssuer jwtIssuer) (*oidc.IDTokenVerifier, error) {
+func newVerifierFromJwtIssuer(o *options.Options, jwtIssuer jwtIssuer) (*internaloidc.IDTokenVerifier, error) {
```
The jwtIssuer does have an Audience, but this is not the only option required. ClientID and ExtraAudiences are also required to build the IDTokenVerificationOptions struct:

https://github.com/AOEpeople/oauth2-proxy/blob/af820cb0a1818a5b1e7f736ac3654ea5a69a4846/pkg/validation/options.go#L405

JoelSpeed replied:
I'd rather we pass these in directly instead of passing the whole options struct, it ties this function too much to the larger configuration struct right now I think, what about something like this as the signature?
```suggestion
func newVerifierFromJwtIssuer(clientID string, extraAudiences []string, jwtIssuer jwtIssuer) (*internaloidc.IDTokenVerifier, error) {
```

JoelSpeed replied:
This comment hasn't been resolved, WDYT?

kschu91 replied:
sorry, I overlooked that. just updated the branch.

kschu91 commented:
@JoelSpeed Thanks for your review. I gone through every of your suggestions and changed them accordingly except the unresolved one. Let me know if you need anything else to merge it.

kschu91 commented:
> @JoelSpeed Thanks for your review. I went through every of your suggestions and changed them accordingly except the unresolved one. Let me know if you need anything else to merge it.

@JoelSpeed ping

JoelSpeed commented on pkg/oidc/verify_test.go:
```diff
@@ -0,0 +1,220 @@
+)
+
+var _ = Describe("Verify", func() {
+	var (
+		ctx context.Context
+	)
+
+	ctx = context.Background()
```
This should be sufficient 🙂 
```suggestion
	ctx := context.Background()
```

czenker replied:
fixed

JoelSpeed reviewed (commented):
I think this is close to being ready to go, please rebase as I noticed some conflicts and see the couple of comments I've left.

CC @NickMeves just want to check you don't have any concerns regarding this feature, seems relatively straight forward to me

NickMeves commented:
> I think this is close to being ready to go, please rebase as I noticed some conflicts and see the couple of comments I've left.
> 
> CC @NickMeves just want to check you don't have any concerns regarding this feature, seems relatively straight forward to me

This is all good with me! I'm a little out of the loop, but I remember the underlying issue from a few months ago. This will be a nice fix to support things like AWS Cognito (though I wish they followed the spec!).

I'll take a quick peek at the diff to check on any docs -- that would be my only interest area.  The `aud` claim and its purpose is frequently a source of confusion to users. I don't want people to disable when they shouldn't as a quick fix 😅

czenker commented:
Hey.

I am a colleague of @kschu91 . He is out of office for a few weeks, so I took over this issue.

I tried my best rebasing on the current `master` and made `golint` and the tests happy. Most of the changes were formatting issues and did not touch the business logic. But I am not sure about the adaptions I needed to make on the Keycloak Test. I seemed to me, the mock was missing the `aud` claims (and a few more), but maybe someone can verify that the mock now represents the reality accordingly. (@pb82, maybe? :smile:)

andytson-inviqa commented:
how would this work in a scenario where you have both authorization flow users and additional client credentials grant clients in Cognito?

if `oidc-audience-claim` was switched to `client_id` then wouldn't the authorization flow users break?

dwestheide commented:
> how would this work in a scenario where you have both authorization flow users and additional client credentials grant clients in Cognito?
> 
> if `oidc-audience-claim` was switched to `client_id` then wouldn't the authorization flow users break?

I can confirm that this is a problem. I have been trying out this branch because we have exactly the scenario described. Once we set `oidc-audience-claim` to `client_id`, the Cognito access tokens for machine clients are working fine. But now the Cognito ID tokens of users that authenticated using the auth code flow are no longer accepted.

My suggestion is to make `oidc-audience-claim` an array of strings. So you could configure the proxy to accept either `client_id` or `aud`, for example. Alternatively, there could always be a fallback check of the `aud` claim if the one configured with `oidc-audience-claim` does not exist.

Any opinions on these two approaches?

andrehaland commented:
> I can confirm that this is a problem. I have been trying out this branch because we have exactly the scenario described. Once we set oidc-audience-claim to client_id, the Cognito access tokens for machine clients are working fine. But now the Cognito ID tokens of users that authenticated using the auth code flow are no longer accepted.
> 
> My suggestion is to make oidc-audience-claim an array of strings. So you could configure the proxy to accept either client_id or aud, for example. Alternatively, there could always be a fallback check of the aud claim if the one configured with oidc-audience-claim does not exist.

I have a similar scenario, and would support both of the suggested approaches

JoelSpeed commented:
@kschu91 @czenker Are you using both client credentials and user ID Tokens? Seems we need to do some work on this PR to allow them both to co-exist

kschu91 commented:
@JoelSpeed @NickMeves Sorry for the late response. I was on parental leave.

However, I changed the implementation to allow specifying multiple audience claims which can co-exist. This makes it possible to specify `aud` and `client_id` which allows to use client access tokens and oidc access tokens from cognito.
@andytson-inviqa andytson-inviqa @andrehaland @dwestheide maybe you can confirm this as well. On our system it works great.

Along with that I found two bugs which I also fixed:
1. The `aud` can be either a string or a list of string. The validation was not able to deal with list of strings which resulted in errors during validation. I fixed that and made the code compatible to both.
2. extra jwt issuers did not work with the new validation. This is also fixed now.

dwestheide commented:
> @andytson-inviqa andytson-inviqa @andrehaland @dwestheide maybe you can confirm this as well. On our system it works great.

Thanks for adapting the suggestion, @kschu91! I will not be able to test this before the 3rd of January, unfortunately. If someone else can do this earlier, do not feel obliged to wait for my confirmation that it works as expected.

kschu91 commented:
@JoelSpeed @NickMeves any news here?

dwestheide commented:
@kschu91 Sorry for taking so long. I finally got around to testing your changes, and I can confirm that I can use both Cognito access tokens and ID tokens now, by configuring both `client_id` and `aud` for the audience claim.

kschu91 commented:
@JoelSpeed @NickMeves firendly reminder, to give this a final review and merge for it

JoelSpeed reviewed (commented):
One outstanding comment otherwise I think this is good, apologies for the delays in reviewing, been a busy few months

JoelSpeed reviewed (approved):
Please add a changelog entry and I'll get this merged, thanks @kschu91

kschu91 commented:
> 

done

JoelSpeed reviewed (approved):
Looks great! Thanks and apologies again for the delays

JoelSpeed commented:
Ahh, looks like there's an issue with the code generation, can you run `make generate` and commit the changes it generates?

kschu91 commented:
> 



> can you run `make generate` and commit the changes it generates

done @JoelSpeed

kschu91 commented:
Awesome, thanks @JoelSpeed

michelpromonet commented:
Hi,

This update will be nice, hope to get it soon in next release.

Best Regards,
Michel.
